File: LeetCode/72EditDistance/EditDistance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        // 2D DP -> Time:O(n*m), Space:O(n*m)
        vector<vector<int>> dp;
        for (int i = 0; i < word1.size()+1; i++) {
            vector<int> temp(word2.size()+1, 0);
            dp.push_back(temp);
        }
        for (int i = word1.size(); i > -1; i--) {
            for (int j = word2.size(); j > -1; j--) {
                if (i == word1.size()) {
                    dp[i][j] = word2.size()-j;
                }
                else if (j == word2.size()) {
                    dp[i][j] = word1.size()-i;
                }
                else if (word1[i] == word2[j]) {
                    dp[i][j] = dp[i+1][j+1];
                }
                else {
                    int tempMin = min(dp[i+1][j], dp[i][j+1]);
                    tempMin = min(tempMin, dp[i+1][j+1]);
                    dp[i][j] = 1 + tempMin;
                }
            }
        }
        return dp[0][0];
    }
};

File: LeetCode/72EditDistance/EditDistance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        # 2D DP -> Time:O(n*m), Space:O(n*m)
        dp = [[0] * (len(word2)+1) for i in range(len(word1)+1)]
        for i in range(len(word1), -1, -1):
            for j in range(len(word2), -1, -1):
                if i == len(word1):
                    dp[i][j] = len(word2)-j
                elif j == len(word2):
                    dp[i][j] = len(word1)-i
                elif word1[i] == word2[j]:
                    dp[i][j] = dp[i+1][j+1]
                else:
                    dp[i][j] = 1 + min(dp[i+1][j], dp[i][j+1], dp[i+1][j+1])
        return dp[0][0]
